lib/auth/idp.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"encoding/xml"
+	"net/url"
+
+	"github.com/beevik/etree"
+	"github.com/crewjam/saml"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth/keystore"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// ProcessSAMLIdPRequest makes a signed SAML response to a SAML auth request.
+//
+//nolint:revive // Because we want this to be IdP.
+func (s *Server) ProcessSAMLIdPRequest(ctx context.Context, req *proto.ProcessSAMLIdPRequestRequest) (*proto.ProcessSAMLIdPRequestResponse, error) {
+	return ProcessSAMLIdPRequest(ctx, s, s.keyStore, req)
+}
+
+// ProcessSAMLIdPRequestClient is a client for retrieving CAs and cluster names, which is
+// necessary for creating SAML responses.
+type ProcessSAMLIdPRequestClient interface {
+	services.AuthorityGetter
+
+	// GetClusterName returns the local cluster name
+	GetClusterName(...services.MarshalOption) (types.ClusterName, error)
+
+	// ListSAMLIdPServiceProviders returns a paginated list of SAML IdP service provider resources.
+	ListSAMLIdPServiceProviders(ctx context.Context, pageSize int, nextToken string) ([]types.SAMLIdPServiceProvider, string, error)
+}
+
+// ProcessSAMLIdPRequest will make a signed SAML response to a SAML auth request.
+// This function is exported for testing in the IdP itself.
+//
+//nolint:revive // Because we want this to be IdP.
+func ProcessSAMLIdPRequest(ctx context.Context, client ProcessSAMLIdPRequestClient, keyStore *keystore.Manager, req *proto.ProcessSAMLIdPRequestRequest) (*proto.ProcessSAMLIdPRequestResponse, error) {
```

Comment:
I notice most of the rest of the implementation is in teleport.e, you could do the same for this with an auth plugin, just a few extra PRs :) Up to you I think

Reply:
Probably the right thing to do. Will do!